### PR TITLE
[WIP] PyTorch 2.0 and torch.compile support

### PR DESCRIPTION
**Pull request by aiddun**
I'm looking into supporting PyTorch >=2.0 to see if the new compiler has any potential speedups - I've found what I think is the root cause of the bug preventing previous upgrade attempts and am seeking advice on the right fix. 

A naive upgrade leads to the same inf checking error in #1610. I believe I've narrowed down the bug to [this line in ray_samplers.py](https://github.com/nerfstudio-project/nerfstudio/blob/3f43e0d4c3885e51f3201cb72fc4cdb262466a29/nerfstudio/model_components/ray_samplers.py#L574), where after the tenth step, pytorch will no longer calculate gradients for the proposal networks, leading to the error. 

My current commit here avoids the error and continues to train by changing `self._steps_since_update > self.update_sched(self._step)` to `self._steps_since_update >= self.update_sched(self._step)` - for the example nerfacto poster command, this clause formerly always resolved to false (1 < 1). I don't know enough about pytorch's inf checks, why the updated flag is the way it currently is in ray_samplers.py, or why upgrading to 2.0 broke this line's usefulness to know if this is a proper solution, however I think it's a start if anyone would like to contribute.

### CODE REVIEW

**Comment by jkulhanek** [comment]
There is another problem with PyTorch 2.0 #1883

**Comment by SauravMaheshkar** [comment]
There is also work being done in https://github.com/nerfstudio-project/nerfstudio/pull/1922

**Comment by jkulhanek** [comment]
The reason why it was broken after upgrading to torch 2.0 is the following:
In the nerfacto model the proposal network does not get updated every training step. It gets updated first 10 steps and then it is driven by a schedule. Changing the strictness of the equality check caused to be updated every iteration - which is wrong. But that was the reason why your code was working in torch 2.0.
The way the proposal network gets updated / or doesn't get updated is done by enabling/disabling gradients to flow through it. Unfortunately, grad-scaler is called on all the optimizers (including the one handling the proposal network) without checking if the gradients were computed. This worked fine in torch 1.13, but doesn't in 2.0. The reason is this:
https://github.com/pytorch/pytorch/pull/92731
... in torch 2.0 grads are reset to None instead of zeros to save memory (which is a breaking change).

In the alternative PR: #1922 the problem is fixed by  skipping grad_scaler.step for optimizers for which the grads weren't computed.

#1922 (for now) provides a compatibility with previous torch and addresses the same problem as this PR (except it does not include model.compile).

I suggest we close this PR?

**Comment by jkulhanek** [comment]
Completer in #1922
